### PR TITLE
implement and fix `wifi_networks` on macOS Big Sur and newer

### LINKED ISSUE #6235: Table columns: last_connected, auto_login, in wifi_networks no longer work
# Bug report

### What operating system and version are you using?

```
osquery> SELECT version, build, platform FROM os_version;
+---------+-------+----------+
| version | build | platform |
+---------+-------+----------+
| 10.15.2 | 19C57 | darwin   |
+---------+-------+----------+
```
### What version of osquery are you using?

```
+---------+
| version |
+---------+
| 4.1.1   |
+---------+
```
### What steps did you take to reproduce the issue?
Run the following query in osqueryi:
```SQL
SELECT * FROM wifi_networks
```

### What did you expect to see?
I expected to see data for the last_connected and auto_login. 
For example the screenshot below contains **LastConnected**:

image

### What did you see instead?
I saw errors and no data produced for either column, despite confirming the presence of data in the corresponding plist (which the table pulls its data from).

```shell
osquery> select network_name, security_type, last_connected, auto_login, captive_portal from wifi_networks WHERE network_name = "Jason's MacBook Pro";
I0211 09:09:10.511170 281939392 dynamic_table_row.cpp:114] Error casting last_connected () to INTEGER
I0211 09:09:10.511199 281939392 dynamic_table_row.cpp:114] Error casting auto_login () to INTEGER
+---------------------+--------------------+----------------+------------+----------------+
| network_name        | security_type      | last_connected | auto_login | captive_portal |
+---------------------+--------------------+----------------+------------+----------------+
| Jason's MacBook Pro | WPA2/WPA3 Personal |                |            | 0              |
+---------------------+--------------------+----------------+------------+----------------+
```

### LINKED ISSUE #7206: `wifi_networks` is broken on recent macOS
# Bug report

The `wifi_networks` table is parsing a plist. Somewhere between Mojave and Big Sur, that plist changed:

```
miniseph:~ seph$ sudo ./osqueryd-4.8.0 -S --verbose "select * from wifi_networks"
I0722 11:55:15.540264 157650432 init.cpp:342] osquery initialized [version=4.8.0]
I0722 11:55:15.540942 157650432 extensions.cpp:453] Could not autoload extensions: Cannot open file for reading: /var/osquery/extensions.load
I0722 11:55:15.541041 157650432 init.cpp:553] An error occurred during extension manager startup: Extensions disabled
I0722 11:55:15.541075 157650432 auto_constructed_tables.cpp:97] Removing stale ATC entries
I0722 11:55:15.541301 157650432 init.cpp:591] Error reading config: config file does not exist: /var/osquery/osquery.conf
I0722 11:55:15.566246 157650432 wifi.mm:163] Key not found : KnownNetworks
I0722 11:55:15.566426 157650432 dispatcher.cpp:149] Thread: 0x109658e00 requesting a stop
I0722 11:55:15.566442 157650432 dispatcher.cpp:122] Thread: 0x109658e00 requesting a join
I0722 11:55:15.566452 157650432 dispatcher.cpp:144] Services and threads have been cleared
```

### PR DESCRIPTION
**Pull request by sharvilshah**
Tested on macOS 11.x and 12.x

Preserves functionality with older macOS versions

Added new columns and marked older ones as hidden:

```
# not selecting ssid and network_name to redact my personal networks
osquery> select security_type, possibly_hidden, roaming_profile, temporarily_disabled, add_reason, added_at, captive_portal, captive_login_date, was_captive_network, auto_join, personal_hotspot from wifi_networks;
+--------------------+-----------------+-----------------+----------------------+------------+------------+----------------+--------------------+---------------------+-----------+------------------+
| security_type      | possibly_hidden | roaming_profile | temporarily_disabled | add_reason | added_at   | captive_portal | captive_login_date | was_captive_network | auto_join | personal_hotspot |
+--------------------+-----------------+-----------------+----------------------+------------+------------+----------------+--------------------+---------------------+-----------+------------------+
| WPA/WPA2 Personal  |                 | Single          | 0                    | WiFi Menu  | 1640288052 | 0              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal | 0               | Single          | 0                    | WiFi Menu  | 1642491317 | 0              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal | 0               | Single          | 0                    | WiFi Menu  | 1624124215 | 0              |                    |                     | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1646023391 | 0              | 1646023408         | 1                   | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1645712825 | 1              | 1645870503         |                     | 0         |                  |
| WPA2/WPA3 Personal | 0               | Single          | 0                    | WiFi Menu  | 1621417270 | 0              |                    |                     | 0         | 1                |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1646211809 | 1              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal |                 | Single          | 0                    | WiFi Menu  | 1621416564 | 0              |                    |                     | 0         | 1                |
| Open               |                 | Dual            | 0                    | WiFi Menu  | 1646211769 |                |                    |                     | 1         |                  |
| WPA/WPA2 Personal  | 0               | Multi           | 0                    | WiFi Menu  | 1616833812 | 0              |                    |                     | 1         |                  |
| Open               | 0               | None            | 0                    | WiFi Menu  | 1646211742 | 1              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal |                 | Single          | 0                    | WiFi Menu  | 1640073772 | 0              |                    |                     | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1637479995 |                |                    |                     | 1         |                  |
| WPA/WPA2 Personal  | 0               | Multi           | 0                    | WiFi Menu  | 1611348583 | 0              |                    |                     | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1637302170 | 0              |                    | 1                   | 1         |                  |
| WPA2/WPA3 Personal | 0               | Multi           | 0                    | WiFi Menu  | 1620156066 | 0              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal | 0               | Multi           | 0                    | WiFi Menu  | 1618212039 | 0              |                    |                     | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1623420518 | 0              | 1623509354         | 1                   | 1         |                  |
| Open               | 0               | Multi           | 0                    | WiFi Menu  | 1637480095 | 1              |                    |                     | 1         |                  |
| WPA/WPA2 Personal  |                 | Single          | 0                    | WiFi Menu  | 1639811737 | 0              |                    |                     | 1         |                  |
| WPA2/WPA3 Personal | 0               | Single          | 0                    | WiFi Menu  | 1639388694 | 0              |                    |                     | 1         |                  |
+--------------------+-----------------+-----------------+----------------------+------------+------------+----------------+--------------------+---------------------+-----------+------------------+
```

Closes #7206
Closes #6235

### CODE REVIEW

File: osquery/tables/networking/darwin/wifi.mm [hunk 1]
```diff
@@ -132,10 +231,11 @@ QueryData genKnownWifiNetworks(QueryContext& context) {
     return {};
   }
 
-  boost::filesystem::path path = kAirPortPreferencesPath;
+  std::string p = isBigSurOrHigher() ? kAirPortPrefPathPostCatalina
+                                     : kAirPortPreferencesPath;
+  boost::filesystem::path path = p;
```
Comment (directionless) [inline osquery/tables/networking/darwin/wifi.mm:236]: Something feels a little weird here. Why assign p? It's just being set to a constant. Why call isBigSurOrHigher twice?

Why not:
```
if isBigSurOrHigher() {
  return parseBigSur(kAirPortPrefPathPostCatalina);
}
```

File: osquery/tables/networking/darwin/wifi_utils.mm [hunk 2]
```diff
@@ -10,11 +10,45 @@
...
+#include <osquery/core/tables.h>
 #include <osquery/tables/networking/darwin/wifi_utils.h>
+#include <osquery/utils/conversions/darwin/cfdata.h>
+#include <osquery/utils/conversions/darwin/cfnumber.h>
+#include <osquery/utils/conversions/darwin/cfstring.h>
 
 namespace osquery {
 namespace tables {
 
+std::string getPropertiesFromDictionary(const CFDictionaryRef& dict,
```
Comment (directionless) [inline osquery/tables/networking/darwin/wifi_utils.mm:22]: This seems fine, we appear to have this code block in a bunch of `osquery/utils/conversions/darwin` which makes me wonder if we should unify it. But, that could not happen in this PR

**Review by directionless — COMMENTED** [review]
Tables work on my m1, though the CI build error means I couldn't test on my intel machine

File: osquery/tables/networking/darwin/wifi.mm (hunk 1 above)
Comment (lhmzhou) [reply osquery/tables/networking/darwin/wifi.mm:236]: Will this function apply for succeeding OS versions like Monterey?

File: osquery/tables/networking/darwin/wifi.mm [hunk 3]
```diff
@@ -124,6 +127,102 @@ void parseNetworks(const CFDictionaryRef& network, QueryData& results) {
   results.push_back(r);
 }
 
+QueryData parseBigSur(const std::string& path) {
+  QueryData results;
+  @autoreleasepool {
+    auto plist = (__bridge CFDictionaryRef)
+        [NSDictionary dictionaryWithContentsOfFile:@(path.c_str())];
+    if (plist == nullptr || CFDictionaryGetCount(plist) == 0) {
+      VLOG(1) << "cannot open plist";
```
[the remark this answers is not in the document]
Comment (lhmzhou) [reply osquery/tables/networking/darwin/wifi.mm:136]: +1

File: osquery/tables/networking/darwin/wifi.mm (hunk 1 above)
Comment (sharvilshah) [reply osquery/tables/networking/darwin/wifi.mm:236]: @lhmzhou Yes it will :)

Comment (sharvilshah) [reply osquery/tables/networking/darwin/wifi.mm:236]: @directionless there is a bit of path validation, and didn't want to repeat the code. FWIW I am working on refactoring and cleaning up this part of code along with wifi_utils in a new PR, that will be cleaner.

File: osquery/tables/networking/darwin/wifi_utils.mm (hunk 2 above)
Comment (sharvilshah) [reply osquery/tables/networking/darwin/wifi_utils.mm:22]: Yes, there is also a bit of IOKit specific parsing. I will revisit when I am doing the refactor and cleanup next

**Comment by sharvilshah** [comment]
Thanks for the review @directionless and @lhmzhou! I do want to land this sooner as this has been broken for a bit. I have manually tested again from macOS 10.13 till now, both on intel and apple silicon. 

I do have a WIP refactor on this, and waiting on the WWDC beta drop to test and PR that.

**Comment by sharvilshah** [comment]
> I think we should try to get this into 5.3.0. Do you think it makes sense to wait for WWDC, or to merge and revisit when there's a beta.

Yeah I think it makes sense to get this in 5.3.0, it's been broken for a bit.

**Review by Smjert — CHANGES_REQUESTED** [review]
Gave a look at the code and tested on Monterey.
Just a minor nit, otherwise looks good!.

**Comment by sharvilshah** [comment]
> docs/readthedocs.org:osquery — Read the Docs build failed!

Closing and reopening

**Comment by sharvilshah** [comment]
Rebased to fix the read the docs failing -- one final re-review please @Smjert @directionless

**Review by Smjert — APPROVED** [review]

**Pull request merged**